akka-stream/src/main/scala/akka/stream/scaladsl/Sink.scala

```diff
@@ -107,37 +103,8 @@ object Sink extends SinkApply {
    * normal end of the stream, or completed with `Failure` if there is a failure signaled in
    * the stream..
    */
-  def foreach[T](f: T ⇒ Unit): Sink[T, Future[Unit]] = {
-
-    def newForeachStage(): (PushStage[T, Unit], Future[Unit]) = {
-      val promise = Promise[Unit]()
-
-      val stage = new PushStage[T, Unit] {
-        override def onPush(elem: T, ctx: Context[Unit]): SyncDirective = {
-          f(elem)
-          ctx.pull()
-        }
-        override def onUpstreamFailure(cause: Throwable, ctx: Context[Unit]): TerminationDirective = {
-          promise.failure(cause)
-          ctx.fail(cause)
-        }
-        override def onUpstreamFinish(ctx: Context[Unit]): TerminationDirective = {
-          promise.success(())
-          ctx.finish()
-        }
-
-        override def decide(cause: Throwable): Supervision.Directive = {
-          // supervision will be implemented by #16916
-          promise.tryFailure(cause)
-          super.decide(cause)
-        }
-      }
-
-      (stage, promise.future)
-    }
-
-    Flow[T].transformMaterializing(newForeachStage).to(Sink.ignore).named("foreachSink")
-  }
+  def foreach[T](f: T ⇒ Unit): Sink[T, Future[Unit]] =
+    Flow[T].map(f).toMat(Sink.ignore)(Keep.right).named("foreachSink")
```

Comment:
I made this change for the simple reason that I really, really, do not like custom stages since the materializer has no clue what the operation does. With map + ignore at least we can have the Materializer see that and optimize accordingly.